decode.go

```diff
@@ -1254,49 +1252,79 @@ func (d *decoder) parseToTime() (tm time.Time, err error) {
 			// Read tag number
 			_, _, tagNum := d.getHead()
 			if tagNum != 0 && tagNum != 1 {
-				d.skip()
-				err = errors.New("cbor: wrong tag number for time.Time, got " + strconv.Itoa(int(tagNum)) + ", expect 0 or 1")
-				return
+				d.skip() // skip tag content
+				return time.Time{}, errors.New("cbor: wrong tag number for time.Time, got " + strconv.Itoa(int(tagNum)) + ", expect 0 or 1")
 			}
 		}
 	} else {
 		if d.dm.timeTag == DecTagRequired {
 			d.skip()
-			err = &UnmarshalTypeError{CBORType: t.String(), GoType: typeTime.String(), errorMsg: "expect CBOR tag value"}
-			return
+			return time.Time{}, &UnmarshalTypeError{CBORType: t.String(), GoType: typeTime.String(), errorMsg: "expect CBOR tag value"}
 		}
 	}
 
-	var content interface{}
-	content, err = d.parse(false)
-	if err != nil {
-		return
-	}
-
-	switch c := content.(type) {
-	case nil:
-		return
-	case uint64:
-		return time.Unix(int64(c), 0), nil
-	case int64:
-		return time.Unix(c, 0), nil
-	case float64:
-		if math.IsNaN(c) || math.IsInf(c, 0) {
-			return
-		}
-		f1, f2 := math.Modf(c)
-		return time.Unix(int64(f1), int64(f2*1e9)), nil
-	case string:
-		tm, err = time.Parse(time.RFC3339, c)
+	switch t := d.nextCBORType(); t {
+	case cborTypeTextString:
+		s, err := d.parseTextString()
 		if err != nil {
-			tm = time.Time{}
-			err = errors.New("cbor: cannot set " + c + " for time.Time: " + err.Error())
-			return
+			return time.Time{}, err
 		}
-		return
+		t, err := time.Parse(time.RFC3339, string(s))
+		if err != nil {
+			return time.Time{}, errors.New("cbor: cannot set " + string(s) + " for time.Time: " + err.Error())
+		}
+		return t, nil
+	case cborTypePositiveInt:
+		_, _, val := d.getHead()
+		if val > math.MaxInt64 {
+			return time.Time{}, &UnmarshalTypeError{
+				CBORType: t.String(),
+				GoType:   typeTime.String(),
+				errorMsg: fmt.Sprintf("%d overflows Go's int64", val),
+			}
+		}
+		return time.Unix(int64(val), 0), nil
+	case cborTypeNegativeInt:
+		_, _, val := d.getHead()
+		if val > math.MaxInt64 {
+			if val == math.MaxUint64 {
+				// Maximum absolute value representable by negative integer is 2^64,
+				// not 2^64-1, so it overflows uint64.
+				return time.Time{}, &UnmarshalTypeError{
+					CBORType: t.String(),
+					GoType:   typeTime.String(),
+					errorMsg: fmt.Sprintf("-18446744073709551616 overflows Go's int64"),
+				}
+			}
+			return time.Time{}, &UnmarshalTypeError{
+				CBORType: t.String(),
+				GoType:   typeTime.String(),
+				errorMsg: fmt.Sprintf("-%d overflows Go's int64", val+1),
+			}
+		}
+		return time.Unix(int64(-1)^int64(val), 0), nil
+	case cborTypePrimitives:
+		_, ai, val := d.getHead()
```

Comment:
We also need to handle CBOR nil and undefined here by returning `time.Time{}` and no error.

Reply:
As far as I can tell, that case only occurs when unrecognized tags in the input are combined with TimeTag=DecTagIgnored. I made a table of the current behavior on master in https://gist.github.com/benluddy/0f43e72070c129363e5faa6afb539921, and some of the DecTagIgnored behaviors look a little asymmetric to me:

1. `0(null)` and `1(null)` are rejected by the built-in tag content type validation, but `3735928559(0(null))` is not.
2. `3735928559(null)` decodes to Time's zero value, but `null` is a no-op.
3. `1(3735928559(null))` is rejected by the built-in tag content type validation, but `3735928559(1(3735928559(null)))` decodes to Time's zero value.

I'm not sure how to reconcile these. Maybe the simplest thing to do is bypass the built-in tag content type validation for tags 0 and 1 when TimeTag is DecTagIgnored, and accept that swallowing tags results in some permissive behavior?

Reply:
Thanks for creating the table of all input combinations!  :heart_eyes:

`DecTagIgnored` option was meant to handle unknown tags.

I agree that `DecTagIgnored` behaviors have some inconsistency.

Some ideas for consideration with `TimeTag=DecTagIgnored`:
- decoding a tagged `null` should be a no-op to be consistent with `null`.
- supported known tags should be validated to be consistent with `0(null)`, etc.

For example:
- `3735928559(0(null))` and `3735928559(1(3735928559(null)))` should be rejected by tag validation.
- `3735928559(null)` should be a no-op as `null`.  

Thoughts?

Thanks again for your insights and contributions!

Reply:
> - `3735928559(0(null))` and `3735928559(1(3735928559(null)))` should be rejected by tag validation.

Although I'm not personally using `DecTagIgnored`, that would be my preference given both examples are well-formed but invalid. There is precedent in other options for decoding potentially-invalid input, like `UTF8DecodeInvalid`, but it's easy to imagine a use case for that one where the input has already been validated or will be validated after decode. After decoding `3735928559(0(null))` with `DecTagIgnored`, a caller would have no signal to detect that there had been an invalid input.

> - `3735928559(null)` should be a no-op as `null`.

This also makes sense to me. The existing behavior for 1-tagged floats with NaN or infinite values (as briefly referenced in https://www.rfc-editor.org/rfc/rfc8949.html#section-3.4.2-6) to support use cases that want to explicitly represent a present-and-zero-valued time. I don't think that is documented -- I'll include documentation updates for that as well as for `TimeTag`.